### PR TITLE
impl(GCS+gRPC): asynchronous `ReadObject()`

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #12116

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12694)

### CODE REVIEW

**Comment by coryan** [comment]
/FYI: @frankyn this should unblock some prototyping for you.

File: google/cloud/storage/examples/storage_async_samples.cc [hunk 1]
```diff
@@ -140,6 +140,40 @@ void InsertObjectVectorVectors(
...
+  namespace gcs_ex = google::cloud::storage_experimental;
+  auto coro =
+      [](gcs_ex::AsyncClient& client, std::string bucket_name,
+         std::string object_name) -> google::cloud::future<std::uint64_t> {
+    auto [reader, token] = (co_await client.ReadObject(std::move(bucket_name),
+                                                       std::move(object_name)))
+                               .value();
+    std::uint64_t count = 0;
+    while (token.valid()) {
+      auto [payload, t_] = (co_await reader.Read(std::move(token))).value();
```
Comment (devbww) [inline google/cloud/storage/examples/storage_async_samples.cc:156]: Why `t_`?

File: google/cloud/storage/tests/async_client_integration_test.cc [hunk 2]
```diff
@@ -127,6 +128,43 @@ TEST_F(AsyncClientIntegrationTest, ComposeObject) {
...
+
+  auto r = async.ReadObject(bucket_name(), object_name).get();
+  ASSERT_STATUS_OK(r);
+  AsyncReader reader;
+  AsyncToken token;
+  std::tie(reader, token) = *std::move(r);
+
+  std::string actual;
+  while (token.valid()) {
+    ReadPayload payload;
```
Comment (devbww) [inline google/cloud/storage/tests/async_client_integration_test.cc:153]: Nit: Move this down a couple of lines?

File: google/cloud/storage/tests/async_client_integration_test.cc [hunk 3]
```diff
@@ -127,6 +128,43 @@ TEST_F(AsyncClientIntegrationTest, ComposeObject) {
...
+    std::tie(payload, token) = *std::move(p);
+    for (auto v : payload.contents()) actual += std::string(v);
+  }
+  EXPECT_EQ(actual, expected_data);
+
+  auto status = async
+                    .DeleteObject(bucket_name(), object_name,
+                                  gcs::Generation(insert->generation()))
+                    .get();
+  EXPECT_STATUS_OK(status);
```
Comment (devbww) [inline google/cloud/storage/tests/async_client_integration_test.cc:165]: Q. Do we need this if we've `ScheduleForDelete(*insert)`?

**Review by devbww — APPROVED** [review]

File: google/cloud/storage/examples/storage_async_samples.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/examples/storage_async_samples.cc:156]: AFAIK `auto [payload, token]` would introduce a "new" `token`, and would not affect the value of `token` in the `while()` loop.

Comment (coryan) [reply google/cloud/storage/examples/storage_async_samples.cc]: Oh, maybe you are asking "why the underscore"?  As opposed to a simple name, no reason. Changed.

File: google/cloud/storage/tests/async_client_integration_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/tests/async_client_integration_test.cc]: Done.

File: google/cloud/storage/tests/async_client_integration_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/tests/async_client_integration_test.cc]: Nope. removed.

File: google/cloud/storage/async_client.h [hunk 4]
```diff
@@ -152,6 +154,43 @@ class AsyncClient {
...
+   *     `IfMetagenerationNotMatch`, `UserProject`, and `AcceptEncoding`.
+   *
+   * @par Idempotency
+   * This is a read-only operation and is always idempotent.
+   *
+   * @par Example
+   * @snippet storage_async_samples.cc read-object
+   */
+  template <typename... Args>
+  future<StatusOr<std::pair<AsyncReader, AsyncToken>>> ReadObject(
```
Comment (frankyn) [inline google/cloud/storage/async_client.h:178]: How do you pass in args?

With ReadObjectRange() i've tried and failed:

```
namespace gcs = google::cloud::storage;
auto options = google::cloud::Options{}.set<gcs::Generation>(0);

client.ReadObject(bucket_name, object_name, 0, 100, options).get();
```

failed with error:

```
google/cloud/storage/examples/storage_async_samples.cc:180:67: error: no matching function for call to 'google::cloud::v2_16::Options::set<google::cloud::storage::v2_16::Generation>(int)'
  180 |       auto options = google::cloud::Options{}.set<gcs::Generation>(0);
      |                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
In file included from ./google/cloud/internal/call_context.h:18,
                 from ./google/cloud/internal/async_read_stream_impl.h:19,
                 from ./google/cloud/completion_queue.h:19,
                 from ./google/cloud/storage/internal/async/connection.h:21,
                 from ./google/cloud/storage/async_client.h:18,
                 from google/cloud/storage/examples/storage_async_samples.cc:19:
./google/cloud/options.h:139:12: note: candidate: 'template<class T> google::cloud::v2_16::Options& google::cloud::v2_16::Options::set(ValueTypeT<T>) &'
  139 |   Options& set(ValueTypeT<T> v) & {
      |            ^~~
./google/cloud/options.h:139:12: note:   template argument deduction/substitution failed:
./google/cloud/options.h: In substitution of 'template<class T> using ValueTypeT = typename T::Type [with T = google::cloud::storage::v2_16::Generation]':
./google/cloud/options.h:139:12:   required by substitution of 'template<class T> google::cloud::v2_16::Options& google::cloud::v2_16::Options::set(ValueTypeT<T>) & [with T = google::cloud::storage::v2_16::Generation]'
google/cloud/storage/examples/storage_async_samples.cc:180:67:   required from here
./google/cloud/options.h:97:9: error: no type named 'Type' in 'struct google::cloud::storage::v2_16::Generation'
   97 |   using ValueTypeT = typename T::Type;
      |         ^~~~~~~~~~
./google/cloud/options.h:158:13: note: candidate: 'template<class T> google::cloud::v2_16::Options&& google::cloud::v2_16::Options::set(ValueTypeT<T>) &&'
  158 |   Options&& set(ValueTypeT<T> v) && {
      |             ^~~
      ```

Comment (coryan) [reply google/cloud/storage/async_client.h:178]: `client.ReadObject(bucket_name, object_name, gcs::Generation(12345));`

Comment (frankyn) [reply google/cloud/storage/async_client.h:178]: gotcha; i tried that as well and see:

```
./google/cloud/storage/async_client.h: In instantiation of 'google::cloud::v2_16::future<google::cloud::storage_experimental::v2_16::AsyncReadObjectRangeResponse> google::cloud::storage_experimental::v2_16::AsyncClient::ReadObject(const std::string&, const std::string&, int64_t, int64_t, RequestOptions&& ...) [with RequestOptions = {google::cloud::storage::v2_16::Generation}; std::string = std::__cxx11::basic_string<char>; int64_t = long int]':
google/cloud/storage/examples/storage_async_samples.cc:181:24:   required from here
./google/cloud/storage/async_client.h:206:42: error: no matching function for call to 'forward<google::cloud::v2_16::Options>(google::cloud::storage::v2_16::Generation&)'
  206 |         SpanOptions(std::forward<Options>(options)...));
      |                     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
```

Comment (coryan) [reply google/cloud/storage/async_client.h:178]: I just wrote an additional example and it compiled for me.  Give me 10m, I will merge this PR and send a second PR with the additional example.

**Pull request merged**